### PR TITLE
8329628: Additional changes after JDK-8329332

### PR DESCRIPTION
**Pull request by vnkozlov**
Additional clean up based on comments (mostly Stefan's) during reviews for [JDK-8329332: Remove CompiledMethod and CodeBlobLayout classes](https://bugs.openjdk.org/browse/JDK-8329332).
- Renamed `CompiledMethod_lock` to `NMethod_lock`.  (I decided to not change JVMTI's `CompiledMethod[Load|Unload]` names).
- Renamed `NMethodIterator::all_blobs` to `NMethodIterator::all`.
- Moved `get_deopt_original_pc()` method from `nmethod` to `frame` class.
- Reverted `CodeCache::find_nmethod()` to previous functionality to allow return `nullptr` and be consistent with `find_blob()`.
- Cleanup some `(nmethod*)` casts.
- Use `for (CodeHeap* heap : *_nmethod_heaps) ` in `CodeCache::nmethod_count()` (it was @stefank suggestion,  I don't know how this C++ magic works). I verified it running with `-XX:+PrintNMethodStatistics`.

Testing tier1-3,xcomp,stress

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329628](https://bugs.openjdk.org/browse/JDK-8329628): Additional changes after JDK-8329332 (**Enhancement** - P4)

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18665/head:pull/18665` \
`$ git checkout pull/18665`

Update a local copy of the PR: \
`$ git checkout pull/18665` \
`$ git pull https://git.openjdk.org/jdk.git pull/18665/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18665`

View PR using the GUI difftool: \
`$ git pr show -t 18665`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18665.diff">https://git.openjdk.org/jdk/pull/18665.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18665#issuecomment-2040794392)

### CODE REVIEW

File: src/hotspot/share/code/codeCache.hpp [hunk 1]
```diff
@@ -337,7 +337,7 @@ class CodeCache : AllStatic {
 // The relaxed iterators only hold the CodeCache_lock across next calls
 template <class T, class Filter, bool is_relaxed> class CodeBlobIterator : public StackObj {
  public:
-  enum LivenessFilter { all_blobs, only_not_unloading };
+  enum LivenessFilter { all, only_not_unloading };
```
Comment (stefank) [inline src/hotspot/share/code/codeCache.hpp:340]: Thanks, I like this. FWIW, the `only` in `only_not_unloading` seems redundant. The code reads well without it, IMHO:

```
  // All nmethods
  NMethodIterator iter(NMethodIterator::all);
```
```
  // Those that are not unloading
  NMethodIterator iter(NMethodIterator::not_unloading);
```

File: src/hotspot/share/code/codeCache.cpp [hunk 2]
```diff
@@ -664,8 +664,8 @@ CodeBlob* CodeCache::find_blob(void* start) {
 
 nmethod* CodeCache::find_nmethod(void* start) {
   CodeBlob* cb = find_blob(start);
-  assert(cb != nullptr, "did not find an nmethod");
-  return cb->as_nmethod();
+  assert(cb == nullptr || cb->is_nmethod(), "did not find an nmethod");
+  return (nmethod*)cb;
```
Comment (stefank) [inline src/hotspot/share/code/codeCache.cpp:668]: There's a call to `find_nmethod` in `ZNMethod::load_oop` that now lacks a null-check. Would you mind adding one?

**Review by stefank — APPROVED** [review]
Looks good. I've added two suggestions. You can choose to make them, handle them as separate issues, or just ignore them. :)

**Review by fisk — COMMENTED** [review]
I was never a fan of the CompiledMethod_lock name, as it is quite general but only protects a very specific thing: the state. With the NMethod_lock it gets slightly more awkward since the concurrent GCs already have an "nmethod lock" in the GC data of nmethods. Could this lock be called NMethodState_lock instead, to more clearly describe what exactly it is about nmethods that it guards?

File: src/hotspot/share/code/codeCache.cpp (hunk 2 above)
Comment (vnkozlov) [reply src/hotspot/share/code/codeCache.cpp:668]: I added asserts in all places where there is no nullptr check.

**Comment by vnkozlov** [comment]
I did all suggested changes.

**Review by stefank — APPROVED** [review]

**Review by fisk — APPROVED** [review]

**Comment by vnkozlov** [comment]
Thank you, @stefank and @fisk, for reviews.

**Pull request closed without merging**